File: java/currencies-example/src/main/java/com/polydyne/api/currencies/example/UpdateCurrency.java
package com.polydyne.api.currencies.example;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import org.json.JSONObject;

import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.JsonNode;
import com.mashape.unirest.http.exceptions.UnirestException;
import com.polydyne.api.currencies.example.model.Currency;

public class UpdateCurrency {

    // Maps 
    private static Map<String, Double> currentExchangeRateMap = new HashMap<String, Double>();
    private static Map<String, Double> currencyCodeAndRateMap = new HashMap<String, Double>();
    private static Map<String, Long> currencyCodeAndIdMap = new HashMap<String, Long>();
    private static List<Currency> currencyList = new ArrayList<Currency>();


    public void getExchangeRates() throws UnirestException {
        // Fetch latest exchange rates from the open exchange rate API
        String url = Main.OPEN_EXCHANGE_RATES_URL + "/latest.json?app_id="+Main.APP_ID;
        if(url!=null){
      
            HttpResponse<JsonNode> rsp = HttpHelper.get(url).asJson();
            String json = rsp.getBody().getObject().get("rates").toString();
            JSONObject jsonObject = new JSONObject(json.trim());

            Iterator<?> keys = jsonObject.keys();

            // Map currency code to exchange rate fetched from open exchange rates API
            while( keys.hasNext() ) {
                String key = (String) keys.next();
                Double value = (Double) jsonObject.getDouble(key);
                currentExchangeRateMap.put(key, value);
            }           
        }
    }

    public void getCurrency() throws UnirestException {

        String url = Main.API_BASE_URL + "/currencies";
        if(url!=null){
            // Request page of currencies
            HttpResponse<Currency[]> rsp = HttpHelper.get(url).asObject(Currency[].class);
            for(Currency currency: rsp.getBody()){
                // Map each currency code to currency rate
                currencyCodeAndRateMap.put(currency.getCurrencyCode(), currency.getExchangeRate());
                // Map each currency code to currencyId
                currencyCodeAndIdMap.put(currency.getCurrencyCode(), currency.getCurrencyId());
            }
        }       
    }

    public void updateCurrencyRates() throws UnirestException {
        
        // Use the maps to find a list of currencies to be updated
        for(Map.Entry<String, Double> entry:currencyCodeAndRateMap.entrySet()){
            
            String key = entry.getKey();
            if(currentExchangeRateMap.containsKey(key)){
                
                if(currentExchangeRateMap.get(key).compareTo(currencyCodeAndRateMap.get(key))!=0){
                    // Add currencies with new exchange rates to the list
                    Currency currency = new Currency();
                    currency.setCurrencyId(currencyCodeAndIdMap.get(key));
                    currency.setCurrencyCode(key);
                    currency.setExchangeRate(currentExchangeRateMap.get(key));
                    
                    currencyList.add(currency);
                }
                else{
                    System.out.println("Skipping "+ entry.getKey());
                }           
            }
            else{
                System.out.println("Ignoring "+ entry.getKey());
            }
        }
        
        // Update currencies using batch API 
       List<Currency> l = CurrencyService.updateCurrencies(currencyList);
       for(int i =0; i<l.size();i++){
           System.out.println(l.get(i).getCurrencyCode() + " "+ l.get(i).getExchangeRate());
       }
    }
}

File: c#/currencies-example/currencies-example/UpdateCurrency.cs
using System;
using RestSharp;
using RestSharp.Authenticators;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Collections;
using System.Collections.Generic;

namespace CurrenciesExample
{
    public class UpdateCurrency
    {
        private static Dictionary<String, Double> currentExchangeRateMap = new Dictionary<String, Double>();
        private static Dictionary<String, Double> currencyCodeAndRateMap = new Dictionary<String, Double>();
        private static Hashtable currencyCodeAndIdMap = new Hashtable();
        private static ArrayList currencyList = new ArrayList();

        public void GetExchangeRates()
        {
            // Fetch latest exchange rates from the open exchange rate API
            String url = Start.OPEN_EXCHANGE_URL + "/latest.json?app_id=" + Start.APP_ID;
            var client = new RestClient();
            client.BaseUrl = new System.Uri(url);
            var request = new RestRequest();
            request.RequestFormat = DataFormat.Json;

            var response = client.Execute(request);
            // Deserialize json and store in a dictionary
            RestSharp.Deserializers.JsonDeserializer json = new RestSharp.Deserializers.JsonDeserializer();
            var jsonObj = json.Deserialize<Dictionary<String, String>>(response);
            // Obtain the jsonObject rates
            var ans = jsonObj["rates"];
            // Parse the rates information to a jsonObject and store in a dictionary mapping currency code to latest exchange rates
            JObject jsonObject = JObject.Parse(ans);
            currentExchangeRateMap = jsonObject.ToObject<Dictionary<String, Double>>();
        }

        public void GetCurrency()
        {
            String url = Start.API_BASE_URL + "/currencies";
            var client = new RestClient();
            client.BaseUrl = new System.Uri(url);
            client.Authenticator = new HttpBasicAuthenticator(Start.USERNAME, Start.PASSWORD);
            var request = new RestRequest();
            request.RequestFormat = DataFormat.Json;
            var response = client.Execute(request);
            String json = response.Content.ToString();
            List<Currency> deserializedCurrency = JsonConvert.DeserializeObject<List<Currency>>(json);
            foreach (Currency currency in deserializedCurrency)
            {
                // Map each currency code to currency rate
                currencyCodeAndRateMap.Add(currency.currencyCode, currency.exchangeRate);
                // Map each currency code to currencyId
                currencyCodeAndIdMap.Add(currency.currencyCode, currency.currencyId);
            }
        }

        public void UpdateCurrencyRates()
        {
            // Use the maps to find a list of currencies to be updates
            foreach (KeyValuePair<String, Double> entry in currencyCodeAndRateMap)
            {
                String key = entry.Key;

                if (currentExchangeRateMap.ContainsKey(key))
                {
                    if (currentExchangeRateMap[key].CompareTo(currencyCodeAndRateMap[key]) != 0)
                    {
                        // Add currencies with new exchange rates
                        Currency currency = new Currency();
                        currency.currencyId = (long)currencyCodeAndIdMap[key];
                        currency.currencyCode = key;
                        currency.exchangeRate = currentExchangeRateMap[key];
                        currencyList.Add(currency);
                    }
                    else
                    {
                        Console.WriteLine("Skipping " + entry.Key);
                    }
                }
                else
                {
                    Console.WriteLine("Ignoring " + entry.Key);
                }
            }

            // Send list of currencies to be updated via batch API
            CurrencyService.UpdateCurrencies(currencyList);
        }
    }

}
